// @antv/l7-core/lib/services/interaction/InteractionService.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));

var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));

var _eventemitter = _interopRequireDefault(require("eventemitter3"));

var _hammerjs = _interopRequireDefault(require("hammerjs"));

var _inversify = require("inversify");

var _types = require("../../types");

var _IInteractionService = require("./IInteractionService");

var _dec, _dec2, _dec3, _class, _class2, _descriptor, _descriptor2, _temp;

var InteractionService = (_dec = (0, _inversify.injectable)(), _dec2 = (0, _inversify.inject)(_types.TYPES.IMapService), _dec3 = (0, _inversify.inject)(_types.TYPES.ILogService), _dec(_class = (_class2 = (_temp = function (_EventEmitter) {
  (0, _inherits2.default)(InteractionService, _EventEmitter);

  function InteractionService() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, InteractionService);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(InteractionService)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _initializerDefineProperty2.default)(_this, "mapService", _descriptor, (0, _assertThisInitialized2.default)(_this));
    (0, _initializerDefineProperty2.default)(_this, "logger", _descriptor2, (0, _assertThisInitialized2.default)(_this));
    _this.hammertime = void 0;

    _this.onHover = function (_ref) {
      var x = _ref.x,
          y = _ref.y,
          type = _ref.type;

      var $containter = _this.mapService.getMapContainer();

      if ($containter) {
        var _$containter$getBound = $containter.getBoundingClientRect(),
            top = _$containter$getBound.top,
            left = _$containter$getBound.left;

        x -= left;
        y -= top;
      }

      var lngLat = _this.mapService.containerToLngLat([x, y]);

      _this.emit(_IInteractionService.InteractionEvent.Hover, {
        x: x,
        y: y,
        lngLat: lngLat,
        type: type
      });
    };

    return _this;
  }

  (0, _createClass2.default)(InteractionService, [{
    key: "init",
    value: function init() {
      this.addEventListenerOnMap();
    }
  }, {
    key: "destroy",
    value: function destroy() {
      if (this.hammertime) {
        this.hammertime.destroy();
      }

      this.removeEventListenerOnMap();
      this.off(_IInteractionService.InteractionEvent.Hover);
    }
  }, {
    key: "triggerHover",
    value: function triggerHover(_ref2) {
      var x = _ref2.x,
          y = _ref2.y;
      this.emit(_IInteractionService.InteractionEvent.Hover, {
        x: x,
        y: y
      });
    }
  }, {
    key: "triggerSelect",
    value: function triggerSelect(id) {
      this.emit(_IInteractionService.InteractionEvent.Select, {
        featureId: id
      });
    }
  }, {
    key: "triggerActive",
    value: function triggerActive(id) {
      this.emit(_IInteractionService.InteractionEvent.Active, {
        featureId: id
      });
    }
  }, {
    key: "addEventListenerOnMap",
    value: function addEventListenerOnMap() {
      var $containter = this.mapService.getMapContainer();

      if ($containter) {
        var hammertime = new _hammerjs.default($containter);
        hammertime.get('pan').set({
          direction: _hammerjs.default.DIRECTION_ALL
        });
        hammertime.get('pinch').set({
          enable: true
        });
        $containter.addEventListener('mousemove', this.onHover);
        $containter.addEventListener('click', this.onHover);
        $containter.addEventListener('mousedown', this.onHover);
        $containter.addEventListener('mouseup', this.onHover);
        $containter.addEventListener('dblclick', this.onHover);
        $containter.addEventListener('contextmenu', this.onHover);
        this.hammertime = hammertime;
        this.logger.debug('add event listeners on canvas');
      }
    }
  }, {
    key: "removeEventListenerOnMap",
    value: function removeEventListenerOnMap() {
      var $containter = this.mapService.getMapContainer();

      if ($containter) {
        $containter.removeEventListener('mousemove', this.onHover);
        $containter.removeEventListener('click', this.onHover);
        $containter.removeEventListener('mousedown', this.onHover);
        $containter.removeEventListener('mouseup', this.onHover);
        $containter.removeEventListener('dblclick', this.onHover);
        $containter.removeEventListener('contextmenu', this.onHover);
      }
    }
  }]);
  return InteractionService;
}(_eventemitter.default), _temp), (_descriptor = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "mapService", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor2 = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "logger", [_dec3], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class2)) || _class);
exports.default = InteractionService;
