// @antv/l7-core/es/services/config/ConfigService.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _applyDecoratedDescriptor from "@babel/runtime/helpers/applyDecoratedDescriptor";
import _merge from "lodash/merge";

var _dec, _dec2, _class, _class2, _temp;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import Ajv from 'ajv';
import { injectable, postConstruct } from 'inversify';
import mapConfigSchema from './mapConfigSchema';
import sceneConfigSchema from './sceneConfigSchema';
import WarnInfo from './warnInfo';
var defaultSceneConfig = {
  id: 'map'
};
var defaultLayerConfig = {
  colors: ['rgb(103,0,31)', 'rgb(178,24,43)', 'rgb(214,96,77)', 'rgb(244,165,130)', 'rgb(253,219,199)', 'rgb(247,247,247)', 'rgb(209,229,240)', 'rgb(146,197,222)', 'rgb(67,147,195)', 'rgb(33,102,172)', 'rgb(5,48,97)'],
  size: 10,
  shape: 'circle',
  scales: {},
  shape2d: ['circle', 'triangle', 'square', 'pentagon', 'hexagon', 'octogon', 'hexagram', 'rhombus', 'vesica'],
  shape3d: ['cylinder', 'triangleColumn', 'hexagonColumn', 'squareColumn'],
  minZoom: -1,
  maxZoom: 24,
  visible: true,
  autoFit: false,
  zIndex: 0,
  blend: 'normal',
  pickedFeatureID: -1,
  enableMultiPassRenderer: true,
  enablePicking: true,
  active: false,
  activeColor: '#2f54eb',
  enableHighlight: false,
  enableSelect: false,
  highlightColor: '#2f54eb',
  selectColor: 'blue',
  enableTAA: false,
  jitterScale: 1,
  enableLighting: false,
  animateOption: {
    enable: false,
    interval: 0.2,
    duration: 4,
    trailLength: 0.15
  }
};
var ajv = new Ajv({
  allErrors: true,
  verbose: true
});
var GlobalConfigService = (_dec = injectable(), _dec2 = postConstruct(), _dec(_class = (_class2 = (_temp = function () {
  function GlobalConfigService() {
    _classCallCheck(this, GlobalConfigService);

    this.sceneConfigCache = {};
    this.sceneConfigValidator = void 0;
    this.mapConfigValidator = void 0;
    this.layerConfigCache = {};
    this.layerConfigValidatorCache = {};
  }

  _createClass(GlobalConfigService, [{
    key: "getSceneConfig",
    value: function getSceneConfig(sceneId) {
      return this.sceneConfigCache[sceneId];
    }
  }, {
    key: "getSceneWarninfo",
    value: function getSceneWarninfo(id) {
      return WarnInfo[id];
    }
  }, {
    key: "setSceneConfig",
    value: function setSceneConfig(sceneId, config) {
      this.sceneConfigCache[sceneId] = _objectSpread({}, defaultSceneConfig, {}, config);
    }
  }, {
    key: "validateSceneConfig",
    value: function validateSceneConfig(data) {
      return this.validate(this.sceneConfigValidator, data);
    }
  }, {
    key: "validateMapConfig",
    value: function validateMapConfig(data) {
      return this.validate(this.mapConfigValidator, data);
    }
  }, {
    key: "getLayerConfig",
    value: function getLayerConfig(layerId) {
      return this.layerConfigCache[layerId];
    }
  }, {
    key: "setLayerConfig",
    value: function setLayerConfig(sceneId, layerId, config) {
      this.layerConfigCache[layerId] = _objectSpread({}, _merge({}, this.sceneConfigCache[sceneId], defaultLayerConfig, config));
    }
  }, {
    key: "registerLayerConfigSchemaValidator",
    value: function registerLayerConfigSchemaValidator(layerName, schema) {
      if (!this.layerConfigValidatorCache[layerName]) {
        this.layerConfigValidatorCache[layerName] = ajv.compile(schema);
      }
    }
  }, {
    key: "validateLayerConfig",
    value: function validateLayerConfig(layerName, data) {
      return this.validate(this.layerConfigValidatorCache[layerName], data);
    }
  }, {
    key: "clean",
    value: function clean() {
      this.sceneConfigCache = {};
      this.layerConfigCache = {};
    }
  }, {
    key: "registerSceneConfigSchemaValidator",
    value: function registerSceneConfigSchemaValidator() {
      this.sceneConfigValidator = ajv.compile(sceneConfigSchema);
      this.mapConfigValidator = ajv.compile(mapConfigSchema);
    }
  }, {
    key: "validate",
    value: function validate(validateFunc, data) {
      if (validateFunc) {
        var valid = validateFunc(data);

        if (!valid) {
          return {
            valid: valid,
            errors: validateFunc.errors,
            errorText: ajv.errorsText(validateFunc.errors)
          };
        }
      }

      return {
        valid: true,
        errors: null,
        errorText: null
      };
    }
  }]);

  return GlobalConfigService;
}(), _temp), (_applyDecoratedDescriptor(_class2.prototype, "registerSceneConfigSchemaValidator", [_dec2], Object.getOwnPropertyDescriptor(_class2.prototype, "registerSceneConfigSchemaValidator"), _class2.prototype)), _class2)) || _class);
export { GlobalConfigService as default };
