// @antv/l7-core/es/services/scene/SceneService.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _initializerDefineProperty from "@babel/runtime/helpers/initializerDefineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _applyDecoratedDescriptor from "@babel/runtime/helpers/applyDecoratedDescriptor";
import _initializerWarningHelper from "@babel/runtime/helpers/initializerWarningHelper";

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _class, _class2, _descriptor, _descriptor2, _descriptor3, _descriptor4, _descriptor5, _descriptor6, _descriptor7, _descriptor8, _descriptor9, _descriptor10, _descriptor11, _descriptor12, _descriptor13, _descriptor14, _descriptor15, _temp;

import { DOM } from '@antv/l7-utils';
import elementResizeEvent, { unbind } from 'element-resize-event';
import { EventEmitter } from 'eventemitter3';
import { inject, injectable } from 'inversify';
import { AsyncParallelHook } from 'tapable';
import { TYPES } from '../../types';
import { createRendererContainer } from '../../utils/dom';
var Scene = (_dec = injectable(), _dec2 = inject(TYPES.SceneID), _dec3 = inject(TYPES.IIconService), _dec4 = inject(TYPES.IFontService), _dec5 = inject(TYPES.IControlService), _dec6 = inject(TYPES.ILogService), _dec7 = inject(TYPES.IGlobalConfigService), _dec8 = inject(TYPES.IMapService), _dec9 = inject(TYPES.ICoordinateSystemService), _dec10 = inject(TYPES.IRendererService), _dec11 = inject(TYPES.ILayerService), _dec12 = inject(TYPES.ICameraService), _dec13 = inject(TYPES.IInteractionService), _dec14 = inject(TYPES.IShaderModuleService), _dec15 = inject(TYPES.IMarkerService), _dec16 = inject(TYPES.IPopupService), _dec(_class = (_class2 = (_temp = function (_EventEmitter) {
  _inherits(Scene, _EventEmitter);

  function Scene() {
    var _this;

    _classCallCheck(this, Scene);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Scene).call(this));

    _initializerDefineProperty(_this, "id", _descriptor, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "iconService", _descriptor2, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "fontService", _descriptor3, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "controlService", _descriptor4, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "logger", _descriptor5, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "configService", _descriptor6, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "map", _descriptor7, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "coordinateSystemService", _descriptor8, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "rendererService", _descriptor9, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "layerService", _descriptor10, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "cameraService", _descriptor11, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "interactionService", _descriptor12, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "shaderModuleService", _descriptor13, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "markerService", _descriptor14, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "popupService", _descriptor15, _assertThisInitialized(_this));

    _this.inited = false;
    _this.initPromise = void 0;
    _this.rendering = false;
    _this.$container = void 0;
    _this.hooks = void 0;

    _this.handleWindowResized = function () {
      _this.emit('resize');

      if (_this.$container) {
        var pixelRatio = window.devicePixelRatio;
        var w = window.innerWidth;
        var h = window.innerHeight;

        if (_this.$container !== document.body) {
          var bounds = _this.$container.getBoundingClientRect();

          w = bounds.right - bounds.left;
          h = bounds.bottom - bounds.top;
        }

        _this.rendererService.viewport({
          x: 0,
          y: 0,
          width: pixelRatio * w,
          height: pixelRatio * h
        });

        DOM.triggerResize();
        _this.coordinateSystemService.needRefresh = true;

        _this.render();
      }
    };

    _this.handleMapCameraChanged = function (viewport) {
      _this.cameraService.update(viewport);

      _this.render();
    };

    _this.hooks = {
      init: new AsyncParallelHook(['config'])
    };
    return _this;
  }

  _createClass(Scene, [{
    key: "init",
    value: function init(sceneConfig) {
      var _this2 = this;

      this.configService.setSceneConfig(this.id, sceneConfig);

      var _this$configService$v = this.configService.validateSceneConfig(this.configService.getSceneConfig(this.id)),
          valid = _this$configService$v.valid,
          errorText = _this$configService$v.errorText;

      if (!valid) {
        this.logger.error(errorText || '');
        return;
      }

      this.shaderModuleService.registerBuiltinModules();
      this.iconService.init();
      this.fontService.init();
      this.hooks.init.tapPromise('initMap', _asyncToGenerator(_regeneratorRuntime.mark(function _callee() {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return new Promise(function (resolve) {
                  _this2.map.onCameraChanged(function (viewport) {
                    _this2.cameraService.init();

                    _this2.cameraService.update(viewport);

                    resolve();
                  });

                  _this2.map.init();
                });

              case 2:
                _this2.map.onCameraChanged(_this2.handleMapCameraChanged);

                _this2.map.addMarkerContainer();

                _this2.markerService.addMarkers();

                _this2.markerService.addMarkerLayers();

                _this2.popupService.initPopup();

                _this2.interactionService.init();

                _this2.logger.debug("map ".concat(_this2.id, " loaded"));

              case 9:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      })));
      this.hooks.init.tapPromise('initRenderer', _asyncToGenerator(_regeneratorRuntime.mark(function _callee2() {
        var $container;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                $container = createRendererContainer(_this2.configService.getSceneConfig(_this2.id).id || '');
                _this2.$container = $container;

                if (!$container) {
                  _context2.next = 9;
                  break;
                }

                _context2.next = 5;
                return _this2.rendererService.init($container);

              case 5:
                elementResizeEvent(_this2.$container, _this2.handleWindowResized);
                window.matchMedia('screen and (-webkit-min-device-pixel-ratio: 1.5)').addListener(_this2.handleWindowResized);
                _context2.next = 10;
                break;

              case 9:
                _this2.logger.error('容器 id 不存在');

              case 10:
                _this2.logger.debug("scene ".concat(_this2.id, " renderer loaded"));

              case 11:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      })));
      this.initPromise = this.hooks.init.promise(this.configService.getSceneConfig(this.id));
      this.render();
    }
  }, {
    key: "addLayer",
    value: function addLayer(layer) {
      this.logger.debug("add layer ".concat(layer.name, " to scene ").concat(this.id));
      this.layerService.add(layer);
      this.render();
    }
  }, {
    key: "render",
    value: function () {
      var _render = _asyncToGenerator(_regeneratorRuntime.mark(function _callee3() {
        return _regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                if (!this.rendering) {
                  _context3.next = 2;
                  break;
                }

                return _context3.abrupt("return");

              case 2:
                this.rendering = true;

                if (this.inited) {
                  _context3.next = 10;
                  break;
                }

                _context3.next = 6;
                return this.initPromise;

              case 6:
                this.logger.info(' render inited');
                this.controlService.addControls();
                this.emit('loaded');
                this.inited = true;

              case 10:
                this.layerService.initLayers();
                this.layerService.renderLayers();
                this.logger.debug("scene ".concat(this.id, " render"));
                this.rendering = false;

              case 14:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function render() {
        return _render.apply(this, arguments);
      }

      return render;
    }()
  }, {
    key: "getSceneContainer",
    value: function getSceneContainer() {
      return this.$container;
    }
  }, {
    key: "ExportMap2Png",
    value: function ExportMap2Png() {
      var _this$$container;

      var renderCanvas = (_this$$container = this.$container) === null || _this$$container === void 0 ? void 0 : _this$$container.getElementsByTagName('canvas')[0];
      this.render();
      var layersPng = renderCanvas === null || renderCanvas === void 0 ? void 0 : renderCanvas.toDataURL();
      return layersPng;
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this.emit('destroy');
      this.inited = false;
      this.layerService.destroy();
      this.interactionService.destroy();
      this.controlService.destroy();
      this.markerService.destroy();
      this.removeAllListeners();
      this.rendererService.destroy();
      this.map.destroy();
      unbind(this.$container, this.handleWindowResized);
      window.matchMedia('screen and (min-resolution: 2dppx)').removeListener(this.handleWindowResized);
    }
  }]);

  return Scene;
}(EventEmitter), _temp), (_descriptor = _applyDecoratedDescriptor(_class2.prototype, "id", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor2 = _applyDecoratedDescriptor(_class2.prototype, "iconService", [_dec3], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor3 = _applyDecoratedDescriptor(_class2.prototype, "fontService", [_dec4], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor4 = _applyDecoratedDescriptor(_class2.prototype, "controlService", [_dec5], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor5 = _applyDecoratedDescriptor(_class2.prototype, "logger", [_dec6], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor6 = _applyDecoratedDescriptor(_class2.prototype, "configService", [_dec7], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor7 = _applyDecoratedDescriptor(_class2.prototype, "map", [_dec8], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor8 = _applyDecoratedDescriptor(_class2.prototype, "coordinateSystemService", [_dec9], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor9 = _applyDecoratedDescriptor(_class2.prototype, "rendererService", [_dec10], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor10 = _applyDecoratedDescriptor(_class2.prototype, "layerService", [_dec11], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor11 = _applyDecoratedDescriptor(_class2.prototype, "cameraService", [_dec12], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor12 = _applyDecoratedDescriptor(_class2.prototype, "interactionService", [_dec13], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor13 = _applyDecoratedDescriptor(_class2.prototype, "shaderModuleService", [_dec14], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor14 = _applyDecoratedDescriptor(_class2.prototype, "markerService", [_dec15], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _descriptor15 = _applyDecoratedDescriptor(_class2.prototype, "popupService", [_dec16], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class2)) || _class);
export { Scene as default };
